File: sdk/typescript/src/internal/local-manifests.ts
import { createHash } from "node:crypto"

import {
  canonicalizeJsonValue,
  type JsonObject,
  type JsonValue,
} from "./jsoncanon"
import {
  compareUtf8,
  validateLocalPackagePath,
} from "./package-graph"

export const LOCAL_MANIFESTS_FORMAT_VERSION = 0 as const

const localManifestsDigestDomain = "helmr.deployment-local-manifests.v0\0"
const maxLocalManifestsSizeBytes = 16777216
const sha256DigestPattern = /^sha256:[0-9a-f]{64}$/

export interface LocalManifestEntry {
  readonly manifestDigest: string
  readonly path: string
}

export interface LocalManifests {
  readonly formatVersion: 0
  readonly entries: readonly LocalManifestEntry[]
}

export function canonicalLocalManifests(manifests: LocalManifests): Uint8Array {
  validateLocalManifests(manifests)
  const canonical = canonicalizeJsonValue(manifests as unknown as JsonValue)
  if (canonical.length > maxLocalManifestsSizeBytes) {
    throw new Error(`local manifests size must be at most ${maxLocalManifestsSizeBytes} bytes`)
  }
  return canonical
}

export function localManifestsDigest(manifests: LocalManifests): Uint8Array {
  const canonical = canonicalLocalManifests(manifests)
  return new Uint8Array(
    createHash("sha256")
      .update(localManifestsDigestDomain)
      .update(canonical)
      .digest(),
  )
}

export function validateLocalManifests(manifests: LocalManifests): void {
  validateLocalManifestsValue(manifests as unknown as JsonValue)
}

function validateLocalManifestsValue(value: JsonValue): LocalManifests {
  const root = requireObject(value, "local manifests")
  requireKeys(root, ["entries", "formatVersion"], "local manifests")
  if (root["formatVersion"] !== LOCAL_MANIFESTS_FORMAT_VERSION) {
    throw new Error(`local manifests formatVersion must be ${LOCAL_MANIFESTS_FORMAT_VERSION}`)
  }

  const entriesValue = requireArray(root["entries"], "local manifests entries")
  if (entriesValue.length === 0) {
    throw new Error("local manifests entries must begin with exactly one root record")
  }
  const entries: LocalManifestEntry[] = []
  const paths = new Set<string>()
  for (const [position, entryValue] of entriesValue.entries()) {
    const object = requireObject(entryValue, `local manifests entries[${position}]`)
    requireKeys(object, ["manifestDigest", "path"], `local manifests entries[${position}]`)
    const path = requireString(object["path"], `local manifests entries[${position}].path`)
    const manifestDigest = requireDigest(
      object["manifestDigest"],
      `local manifests entries[${position}].manifestDigest`,
    )
    if (position === 0) {
      if (path !== ".") {
        throw new Error("local manifests entries must begin with exactly one root record")
      }
    } else {
      if (path === ".") {
        throw new Error("only the first local manifest may use root path .")
      }
      validateLocalPackagePath(path, `local manifests entries[${position}].path`)
      if (position > 1 && compareUtf8(entries[position - 1]!.path, path) >= 0) {
        throw new Error(`local manifests entries are not in canonical path order at position ${position}`)
      }
    }
    if (paths.has(path)) {
      throw new Error(`local manifests entries contains duplicate path ${JSON.stringify(path)}`)
    }
    for (let separator = path.indexOf("/"); separator >= 0;) {
      const ancestor = path.slice(0, separator)
      if (paths.has(ancestor)) {
        throw new Error(
          `local manifest paths ${JSON.stringify(ancestor)} and ${JSON.stringify(path)} overlap`,
        )
      }
      separator = path.indexOf("/", separator + 1)
    }
    paths.add(path)
    entries.push({ manifestDigest, path })
  }
  return { formatVersion: LOCAL_MANIFESTS_FORMAT_VERSION, entries }
}

function requireObject(value: JsonValue | undefined, label: string): JsonObject {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw new Error(`${label} must be an object`)
  }
  return value as JsonObject
}

function requireArray(value: JsonValue | undefined, label: string): readonly JsonValue[] {
  if (!Array.isArray(value)) {
    throw new Error(`${label} must be an array`)
  }
  return value
}

function requireString(value: JsonValue | undefined, label: string): string {
  if (typeof value !== "string") {
    throw new Error(`${label} must be a string`)
  }
  return value
}

function requireDigest(value: JsonValue | undefined, label: string): string {
  const digest = requireString(value, label)
  if (!sha256DigestPattern.test(digest)) {
    throw new Error(`${label} must be a lowercase SHA-256 digest`)
  }
  return digest
}

function requireKeys(value: JsonObject, expected: readonly string[], label: string): void {
  const keys = Object.keys(value).sort()
  if (keys.length !== expected.length || keys.some((key, index) => key !== expected[index])) {
    throw new Error(`${label} has unknown or missing members`)
  }
}

File: internal/deployment/local_manifests.go
package deployment

import (
	"bytes"
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"strings"

	"github.com/helmrdotdev/helmr/internal/jsoncanon"
)

const (
	LocalManifestsFormatVersion = 0
	localManifestsDigestDomain  = "helmr.deployment-local-manifests.v0\x00"
)

type LocalManifestEntry struct {
	ManifestDigest string `json:"manifestDigest"`
	Path           string `json:"path"`
}

type LocalManifests struct {
	FormatVersion int                  `json:"formatVersion"`
	Entries       []LocalManifestEntry `json:"entries"`
}

func CanonicalLocalManifests(manifests LocalManifests) ([]byte, error) {
	if err := ValidateLocalManifests(manifests); err != nil {
		return nil, err
	}
	raw, err := json.Marshal(manifests)
	if err != nil {
		return nil, fmt.Errorf("encode local manifests: %w", err)
	}
	canonical, err := jsoncanon.Transform(raw)
	if err != nil {
		return nil, fmt.Errorf("canonicalize local manifests: %w", err)
	}
	if len(canonical) > int(maxProgramFileSizeBytes) {
		return nil, fmt.Errorf("local manifests size is outside [1,%d]", maxProgramFileSizeBytes)
	}
	return canonical, nil
}

func LocalManifestsDigest(manifests LocalManifests) ([sha256.Size]byte, error) {
	canonical, err := CanonicalLocalManifests(manifests)
	if err != nil {
		return [sha256.Size]byte{}, err
	}
	return domainDigest(localManifestsDigestDomain, canonical), nil
}

func ValidateLocalManifests(manifests LocalManifests) error {
	if manifests.FormatVersion != LocalManifestsFormatVersion {
		return fmt.Errorf(
			"local manifests formatVersion = %d, want %d",
			manifests.FormatVersion,
			LocalManifestsFormatVersion,
		)
	}
	if manifests.Entries == nil {
		return fmt.Errorf("local manifests entries must be an array")
	}
	if len(manifests.Entries) == 0 || manifests.Entries[0].Path != "." {
		return fmt.Errorf("local manifests entries must begin with exactly one root record")
	}

	paths := make(map[string]struct{}, len(manifests.Entries))
	for position, entry := range manifests.Entries {
		if !sha256DigestPattern.MatchString(entry.ManifestDigest) {
			return fmt.Errorf(
				"local manifests entries[%d].manifestDigest is not a lowercase SHA-256 digest",
				position,
			)
		}
		if position == 0 {
			if entry.Path != "." {
				return fmt.Errorf("local manifests root path = %q, want %q", entry.Path, ".")
			}
		} else {
			if entry.Path == "." {
				return fmt.Errorf("root path may only appear in the first local manifest")
			}
			if err := validatePackagePath(entry.Path, programMountPath, true); err != nil {
				return fmt.Errorf("local manifests entries[%d].path: %w", position, err)
			}
			if position > 1 && bytes.Compare(
				[]byte(manifests.Entries[position-1].Path),
				[]byte(entry.Path),
			) >= 0 {
				return fmt.Errorf("local manifests entries are not in canonical path order at position %d", position)
			}
		}
		if _, exists := paths[entry.Path]; exists {
			return fmt.Errorf("local manifests entries contains duplicate path %q", entry.Path)
		}
		for separator := strings.IndexByte(entry.Path, '/'); separator >= 0; {
			ancestor := entry.Path[:separator]
			if _, exists := paths[ancestor]; exists {
				return fmt.Errorf("local manifest paths %q and %q overlap", ancestor, entry.Path)
			}
			next := strings.IndexByte(entry.Path[separator+1:], '/')
			if next < 0 {
				break
			}
			separator += next + 1
		}
		paths[entry.Path] = struct{}{}
	}
	return nil
}
